aslam_cv_common/include/aslam/common/yaml-serialization-eigen.h

```diff
@@ -22,107 +22,142 @@ struct convert<Eigen::Matrix<Scalar_, A_, B_, C_, D_, E_> > {
     CHECK_GT(rows, 0);
     CHECK_GT(cols, 0);
     for (IndexType i = 0; i < rows; ++i) {
-      for (IndexType j = 0; j < cols; ++j) {
-        node["data"].push_back(M(i, j));
+      if (cols > 1) {
+        YAML::Emitter ith_row;
+        ith_row << YAML::Flow;
+        ith_row << YAML::BeginSeq;
+        for (IndexType j = 0; j < cols; ++j) {
+          ith_row << M(i, j);
+        }
+        ith_row << YAML::EndSeq;
+        CHECK(ith_row.good()) << "Emitter error: " << ith_row.GetLastError();
+        node["data"].push_back(YAML::Load(ith_row.c_str()));
+      } else {
+        node["data"].push_back(M(i, 0));
       }
     }
     return node;
   }
 
   template <class Scalar, int A, int B, int C, int D, int E>
-  static bool decode(const Node& node,
-                     Eigen::Matrix<Scalar, A, B, C, D, E>& M) {
-    if(!node.IsMap()) {
-      LOG(ERROR) << "Unable to get parse the matrix because the node is not a map.";
-      return false;
-    }
-
+  static bool decodeEigenMatrixData(
+      const Node& node, const int rows, const int cols,
+      Eigen::Matrix<Scalar, A, B, C, D, E>& M) {
     typedef typename Eigen::Matrix<Scalar, A, B, C, D, E>::Index IndexType;
-    IndexType rows = node["rows"].as<IndexType>();
-    IndexType cols = node["cols"].as<IndexType>();
-
-    if(rows != A || cols != B) {
-      LOG(ERROR) << "The matrix is the wrong size (rows, cols). Wanted: (" << A << ","
-          << B << "), got (" << rows << ", " << cols << ")";
-      return false;
-    }
-
-    size_t expected_size = M.rows() * M.cols();
-    if (!node["data"].IsSequence()) {
+    if (!node.IsSequence()) {
       LOG(ERROR) << "The matrix data is not a sequence.";
       return false;
     }
-    if(node["data"].size() != expected_size) {
-      LOG(ERROR) << "The data sequence is the wrong size. Wanted: " << expected_size <<
-          ", got: " << node["data"].size();
-      return false;
+
+    if (rows == 0 && cols == 0) {
+      return true;
     }
+    YAML::const_iterator it = node.begin();
+    YAML::const_iterator it_end = node.end();
 
-    YAML::const_iterator it = node["data"].begin();
-    YAML::const_iterator it_end = node["data"].end();
-    if (rows > 0 && cols > 0) {
+    // If the 2D matrix is stored as a linear array:
+    if (static_cast<int>(node.size()) == rows * cols) {
```

Comment:
@mfehr, you did this part, was there any reason the things are declared as int, also for the template parameters?

Reply:
this is just to prevent the signed unsigned comparison warning, eigen cols and rows returns an int. We could check them and cast them to int, but I think this is the lazier, but equally safe way to do it.